app/pagestack.go

```diff
@@ -0,0 +1,159 @@
+// the page that will be displayed next via the page.OnNavigatedTo() method.
+func (pageStack *PageStack) Pop() bool {
+	pageStack.mtx.Lock()
+	defer pageStack.mtx.Unlock()
+
+	l := len(pageStack.pages)
+	if l == 0 {
+		return false
+	}
+
+	pageToPop := pageStack.pages[l-1]
+	pageToPop.OnNavigatedFrom()
+	if closeablePage, ok := pageToPop.(Closable); ok {
+		closeablePage.OnClosed()
+	}
+
+	pageStack.pages = pageStack.pages[:l-1]
+	if l > 1 {
+		pageStack.pages[l-2].OnNavigatedTo() // get previous page ready for display
+	}
+	pageStack.debugLog()
+	return true
+}
+
+// PopAfter removes all pages from the top of the stack until (and excluding) a
+// specific page. The matcher parameter should return true for the page that
+// should be excluded. If the matcher never matches a page to exclude, no page
+// will be popped. If any page is popped, the page's OnNavigatedFrom() and if
+// supported, the OnClosed() methods will be called to signal that the page has
+// been removed from the display and will never be re-displayed. The page to be
+// displayed will receive an about-to-display signal via the OnNavigatedTo()
+// method.
+func (pageStack *PageStack) PopAfter(matcher func(Page) bool) bool {
+	retainPageIndex := -1
+	for i := len(pageStack.pages) - 1; i >= 0; i-- {
+		if matcher(pageStack.pages[i]) {
+			retainPageIndex = i
+			break
+		}
+	}
+	if retainPageIndex == -1 {
+		return false
+	}
+
+	popped := pageStack.pages[retainPageIndex+1:] // pop pages after the retainPageIndex
+	for _, poppedPage := range popped {
+		poppedPage.OnNavigatedFrom()
+		if closeablePage, ok := poppedPage.(Closable); ok {
+			closeablePage.OnClosed()
+		}
+	}
+
+	pageStack.pages = pageStack.pages[:retainPageIndex+1] // keep pages from index 0 up till retainPageIndex
+	pageStack.pages[retainPageIndex].OnNavigatedTo()
+	pageStack.debugLog()
+	return true
+}
+
+// Reset pops all pages in the stack and creates a new stack with the specified
+// pages as root. Each popped page's OnNavigatedFrom() and if supported, the
+// OnClosed() methods will be called to signal that the page has been removed
+// from the display and will never be re-displayed. If there are new pages to
+// display, the top page is readied for display via the its OnNavigatedTo()
+// method.
+func (pageStack *PageStack) Reset(newPages ...Page) {
+	pageStack.mtx.Lock()
+	defer pageStack.mtx.Unlock()
+
+	// Close all the pages in the current stack before resetting.
+	for _, existingPage := range pageStack.pages {
+		existingPage.OnNavigatedFrom() // Rename to Close()
+	}
+
+	pageStack.pages = newPages
+	if l := len(newPages); l > 0 {
+		pageStack.pages[l-1].OnNavigatedTo()
+	}
+	pageStack.debugLog()
+}
+
+func (pageStack *PageStack) debugLog() {
+	if l := len(pageStack.pages); l > 0 {
+		fmt.Printf("%s | page to be displayed: %s | stack: %d \n", pageStack.name, pageStack.pages[l-1].ID(), l)
+	} else {
+		fmt.Printf("%s | empty page stack \n", pageStack.name)
```

Comment:
rather than use `fmt`, why not use `logger`?

Reply:
debugLog was for test purposes, will remove in my next PR.